tracer/src/Datadog.Trace/Propagators/W3CBaggagePropagator.cs

```diff
@@ -124,73 +131,55 @@ internal static string Encode(string source, HashSet<char> charsToEncode)
 
             // slice the buffer down to the actual bytes written
             var stackBytes = stackBuffer[..byteCount];
-            return EncodeBytes(source, stackBytes, charsToEncode);
+            EncodeBytesAndAppend(sb, stackBytes, charsToEncode);
+            return;
         }
 #endif
 
         // rent a buffer for the UTF-8 bytes
         var buffer = ArrayPool<byte>.Shared.Rent(maxByteCount);
-        string result;
 
         try
         {
             byteCount = Encoding.UTF8.GetBytes(source, 0, source.Length, buffer, 0);
 
             // slice the buffer down to the actual bytes written
             var bytes = buffer.AsSpan(0, byteCount);
-            result = EncodeBytes(source, bytes, charsToEncode);
+            EncodeBytesAndAppend(sb, bytes, charsToEncode);
         }
         finally
         {
             ArrayPool<byte>.Shared.Return(buffer);
         }
+    }
 
-        return result;
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        static bool ByteRequiresEncoding(byte b, HashSet<char> charsToEncode)
-        {
-            return b < 0x20 || b > 0x7E || char.IsWhiteSpace((char)b) || charsToEncode.Contains((char)b);
-        }
-
-        static bool AnyByteRequiresEncoding(Span<byte> bytes, HashSet<char> charsToEncode)
+    private static void EncodeBytesAndAppend(StringBuilder sb, Span<byte> bytes, HashSet<char> charsToEncode)
+    {
+        foreach (var b in bytes)
         {
-            foreach (var b in bytes)
+            if (b < 0x20 || b > 0x7E || char.IsWhiteSpace((char)b) || charsToEncode.Contains((char)b))
             {
-                if (ByteRequiresEncoding(b, charsToEncode))
-                {
-                    return true;
-                }
+                // encode byte as '%XX'
+                sb.Append($"%{b:X2}");
```

Comment:
[How far do you want to go](https://learn.microsoft.com/en-us/dotnet/api/system.ispanformattable.tryformat?view=net-8.0#system-ispanformattable-tryformat(system-span((system-char))-system-int32@-system-readonlyspan((system-char))-system-iformatprovider)) in the pursuit of fewer allocations here 😄 


```suggestion
#if NET6_0_OR_GREATER
                Span<char> buffer = stackalloc char[2];
                if (b.TryFormat(buffer, out var chars, format: "X2") && chars == 2)
                {
                    sb.Append('%').Append(buffer);
                }
                else
                {
                    // fallback
                    sb.Append($"%{b:X2}");
                }
#else
                sb.Append($"%{b:X2}");
#endif                
```

Reply:
Nice, thanks!

Reply:
I cleaned this up even more by using our own `HexString` to convert the byte to hex and doing it on the stack on .NET 3.1+ or in a rented array otherwise.